File: js/iterators_generators/doklad/python/custom_iterator.py
class Obj():  # Итерируемый объект
  def __init__(self, word):
    self.word = word
 
  def __iter__(self):
    return Iterator(self.word)
 
class Iterator:   # Итератор
  def __init__(self, word):
    self.word = word
    self.index = 0
 
  def __next__(self):
    try:
      letter = self.word[self.index]
      self.index += 1
      return letter  
    except IndexError:
      raise StopIteration()    
 
  def __iter__(self):
    return self
 
 
obj = Obj('sergey')
it = iter(obj)
 
print(it.__next__())
print(it.__next__())
print(it.__next__())
print(it.__next__())
print(it.__next__())
print(it.__next__())
 
print(it.__iter__())

File: js/iterators_generators/doklad/javascript/custom_iterator.js
class IterObj {
  constructor(word) {
    this.word = word;
  }

  iter() { return new Iterator(this.word); }
}


class Iterator {
  constructor(word) {
    this.word = word;
    this.index = 0;
  }

  next() {
    try {
      let letter = this.word[this.index];
      this.index += 1;
      return letter;
    } catch (err) {
      throw "StopIteration";
    }
  }

  iter() { return this; }
}

const iterObj = new IterObj('sergey');
it = iterObj.iter();

console.log(it.next());
console.log(it.next());
console.log(it.next());
console.log(it.next());
console.log(it.next());
console.log(it.next());

console.log('iter obj', IterObj);
console.log('iterator', it);


// OUTPUT:

// s
// e
// r
// g
// e
// y

// iter obj class IterObj {
//   constructor(word) {
//     this.word = word;
//   }
//   iter() { return new Iterator(this.word); }
// }

// iterator Iterator { word: 'sergey', index: 6 }